File: Libc++/Libc++/OSNetwork/UDP.cpp
//
//  UDP.cpp
//  Libc++
//
//  Created by Adam Kopeć on 8/18/17.
//  Copyright © 2017 Adam Kopeć. All rights reserved.
//

#include <OSRuntime.hpp>
#include <string.h>
#include "UDP.hpp"

// Net defs

// hton = Host To Network
static inline short htons_(short v) {
    __asm__("xchg %h0, %b0" : "+Q"(v));
    return (v);
}
// ntoh = Network To Host
#define ntohs_(v) htons(v)
#define ntohl_(v) htonl(v)

void
UDP::send(const void* data, uint32_t length, uint16_t sourcePort, uint16_t destinationPort, IP4_t destinationIP) {
    Packet* packet = (Packet*)OSRuntime::OSMalloc(sizeof(Packet)+length);
    memcpy(packet+1, data, length);
    
    packet->sourcePort = htons_(sourcePort);
    packet->destinationPort = htons_(destinationPort);
    packet->length = htons_(length + sizeof(Packet));
    packet->checksum = 0;
    IP::send((void*)packet, sizeof(Packet)+length, destinationIP, 17, 0);
    OSRuntime::OSFree(packet);
}

File: Kernel/Swift Extensions/Frameworks/Networking/Networking/UDP.swift
//
//  UDP.swift
//  Networking
//
//  Created by Adam Kopeć on 3/3/18.
//  Copyright © 2018 Adam Kopeć. All rights reserved.
//

public struct UDP {
    let networkDevice: NetworkModule
    public init(device: NetworkModule) {
        networkDevice = device
    }
    
    public func send(data: UnsafeRawPointer, length: Int, sourcePort: Int, destinationPort: Int, destinationIP: IPv4) {
        let packet = UnsafeMutablePointer<UDPPacket>(OpaquePointer(malloc(MemoryLayout<UDPPacket>.size + length)))!
        memcpy(packet.advanced(by: 1), data, length)
        
        packet.pointee.sourcePort = UInt16(htons_(Int16(sourcePort)))
        packet.pointee.destinationPort = UInt16(htons_(Int16(destinationPort)))
        packet.pointee.length = UInt16(htons_(Int16(MemoryLayout<UDPPacket>.size + length)))
        packet.pointee.checksum = 0
        
        networkDevice.ip.send(data: packet, length: length + MemoryLayout<UDPPacket>.size, IP: destinationIP, proto: 17, offloading: 0)
        UnsafeRawPointer(packet).deallocate()
    }
}
